shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+-- when we add two 'Value's the resulting 'Value' has, for each asset,
+-- the sum of the quantities of /that particular/ asset in the argument
+-- 'Value'. The effect of this is that the assets in the 'Value' are "independent",
+-- and are operated on separately.
+--
+-- Whenever we need to get the quantity of an asset in a 'Value' where there
+-- is no explicit quantity of that asset in the 'Value', then the quantity is
+-- taken to be zero.
+--
+-- See note [Assets] for more details.
+
+-- | Value type
+data Value crypto = Value
+  { val :: Map  (PolicyID crypto) (Map AssetID Quantity) }
+  deriving (Show, Generic)
+
+instance NoUnexpectedThunks (Value crypto)
+
+-- data ValueBSType = ValueBSType (Map ByteString (Map ByteString Quantity))
+--   deriving (Show, Eq, Generic)
+--
+-- instance NoUnexpectedThunks ValueBSType
+--
+-- -- | make a crypto-free Value type
+-- -- TODO this is a hack!
+-- toValBST :: Value crypto -> ValueBSType
+-- toValBST (Value v) = ValueBSType $ fromList $ fmap (\(cid, tkns) -> (pack $ show cid, tkns)) (toList v)
+
+-- | compact representation of Value
+data CompactValue crypto = AdaOnly Coin | MixValue (Value crypto)
+  deriving (Show, Eq, Generic)
+
+instance NoUnexpectedThunks (CompactValue crypto)
+
+
+-- get the quantities of the tokens of a value term
+getQs :: Value crypto -> [Quantity]
+getQs (Value v) = fmap snd (concat $ fmap toList (elems v))
+
+-- zero value
+zeroV :: Value crypto
+zeroV = Value empty
+
+-- | make a value out of a coin
+coinToValue :: Coin -> Value crypto
+coinToValue (Coin c) = Value $ singleton adaID (singleton adaToken (Quantity c))
+
+-- | make a value out of a coin
+getAdaAmount :: Value crypto -> Coin
+getAdaAmount (Value v) = Coin $ c
+  where
+    Quantity c = foldl (+) (Quantity 0) (getQs $ Value $ filterWithKey (\k _ -> k == adaID) v)
+
+-- | policy ID of Ada
+adaID :: PolicyID crypto
+adaID = PolicyID $ ScriptHash $ coerce ("" :: Hash ShortHash (Script crypto))
+
+-- | asset ID of Ada
+adaToken :: AssetID
+adaToken = AssetID $ pack "Ada"
+
+valueToCompactValue :: Value crypto -> CompactValue crypto
+valueToCompactValue vl
+  | keys v == [adaID] = AdaOnly $ getAdaAmount vl
+  | otherwise         = MixValue $ uniqueAdaToken vl
+    where
+      (Value v) = vl
+
+-- | make a Value term where the only token name within the ada asset group is adaToken
+-- no 0-valued
+uniqueAdaToken :: Value crypto -> Value crypto
+uniqueAdaToken (Value v) = Value $ insert adaID
+  (singleton adaToken $ foldl (+) 0
+  (fmap snd $ concat $ fmap toList (elems (filterWithKey (\k _ -> k == adaID) v))) )
+  (removeZeros $ filterWithKey (\k _ -> k /= adaID) v)
+  where
+    removeZeros vv = fmap (filterWithKey (\_ q -> q /= (Quantity 0))) vv
+
+compactValueToValue :: CompactValue crypto -> Value crypto
+compactValueToValue (AdaOnly c)  = coinToValue c
+compactValueToValue (MixValue v) = v
+
+instance Eq (Value crypto) where
+    {-# INLINABLE (==) #-}
```

Comment:
You probably don't need these pragmas, I'm guessing you copied them from Plutus?